### ISSUE TITLE
typescript effects need a action as a parameter

### ISSUE THREAD

**Issue by cnyballk** [rematch/rematch#478]
Hi. I need some help. I introduced typescript in react and rematch, but I ran into some problems that looked like this.
![image](https://user-images.githubusercontent.com/25214319/42864508-4c6e445c-8a99-11e8-8aa7-25ad4221e524.png)
In the definition,
![image](https://user-images.githubusercontent.com/25214319/42864532-63a18e4a-8a99-11e8-8453-0f5f3471055e.png)
It reminds me that I need a action as a parameter, but I'm pretty clear I don't need one
![image](https://user-images.githubusercontent.com/25214319/42864559-7425c6be-8a99-11e8-9e89-35478de41502.png)
Now I don't know what I'm gonna do. I don't want to use as any. Can you help me?

**Comment by cnyballk** [comment]
This is my environment.
 "@rematch/core": "^1.0.0-beta.4",

**Comment by ShMcK** [comment]
In the case of `models` are you handling them in a similar way to this link?

https://github.com/rematch/rematch/blob/master/examples/ts/count/src/store.ts#L7

**Comment by cnyballk** [comment]
Yes, I did.

**Comment by Izhaki** [comment]
Running into the same issue with:

```json
  "dependencies": {
    "@rematch/core": "^1.0.0-beta.4"
  },
  "devDependencies": {
    "typescript": "3.0.3"
  }
```

models.js:

```typescript
import { createModel } from '@rematch/core';

export const auth = createModel({
  state: 0, // initial state
  effects: dispatch => ({
    userAlreadyLoggedIn() {
      console.log('SIM: userAlreadyLoggedIn');
    },
  }),
});
```

index.js:

```typescript
import { init, RematchDispatch } from '@rematch/core';
import * as models from './models';

type IModels = typeof models;

const store = init({
  models,
});

const dispatch: RematchDispatch<IModels> = store.dispatch;

dispatch.auth.userAlreadyLoggedIn();
// [ts] Expected 1 arguments, but got 0.
// (method) userAlreadyLoggedIn(action: Action<any, any>): Promise<Dispatch<Action<any, any>>> (+1 overload)

```

**Comment by derrickbeining** [comment]
I'm having the same problem.

```
  "dependencies": {
    "@rematch/core": "^1.0.0-beta.5"
  },
  "devDependencies": {
    "typescript": "3.1.1"
  }
```

I've been trying to decipher the [typings.d.ts](https://github.com/rematch/rematch/blob/master/src/typings.d.ts), but I'm pretty new to TypeScript, so I'm not entirely sure if I'm on the right track.

The type signature for `effects` ultimately gets its type from the generic type `RematchDispatcherAsync` ([typings.d.ts L75-80](https://github.com/rematch/rematch/blob/7554420d279fc1cc6095cc7a9d84030814325b66/src/typings.d.ts#L75)), which is called by `ExtractRematchDispatcherAsyncFromEffect<E>` ([typings.d.ts L17](https://github.com/rematch/rematch/blob/1.0.0-beta.5/src/typings.d.ts#L17)):

```ts
export type RematchDispatcherAsync<P = void, M = void> =
  ((action: Action<P, M>) => Promise<Redux.Dispatch<Action<P, M>>>) &
  ((action: Action<P, void>) => Promise<Redux.Dispatch<Action<P, void>>>) &
  (P extends void ? ((...args: any[]) => Promise<Action<any, any>>) :
    M extends void ? ((payload: P) => Promise<Action<P, void>>) :
    (payload: P, meta: M) => Promise<Action<P, M>>)
```

I'm not familiar with how intersections with function types works, so the above type is a bit of a mystery to me. They obviously all return a `Promise<Action>`, which doesn't seem right. `reducers` return an action, but as far as I know, `effects` return a `Promise` for whatever you return from them. I see that @brunolemos did a bit of work in this area a while ago (thanks btw!). Any chance you might have a hunch what's going on here?

**Comment by derrickbeining** [comment]
I managed to get rid of the type error I was getting (see #526) AND get the correct function signature for `distpatch.model.effectName` by changing `types.d.ts` so that `ExtractRematchDispatchersFromEffectsObject<effects>` didn't use `ExtractRematchDispatcherAsyncFromEffect<E>` and `RematchDispatcherAsync<P, M>` and instead just used the lookup type.

I.e. went from this:

```ts
export type ExtractRematchDispatchersFromEffectsObject<effects extends ModelEffects<any>> = {
  [effectKey in keyof effects]: ExtractRematchDispatcherAsyncFromEffect<effects[effectKey]>
}
```

to this:

```ts
export type ExtractRematchDispatchersFromEffectsObject<effects extends ModelEffects<any>> = {
  [effectKey in keyof effects]: effects[effectKey]
}
```

Just need to know... will this break something? I cloned `rematch` and made the change to `typings.d.ts` and it still compiled, but maybe I'm missing something.
